xchange-stream-gateio/src/main/java/GateioStreamingService.java

```diff
@@ -0,0 +1,144 @@
+    final ObjectMapper mapper = StreamingObjectMapperHelper.getObjectMapper();
+
+    return subscribeChannel(channelName, currencyPair, args)
+        .map(
+            msg -> {
+              switch (channelName) {
+                case SPOT_ORDERBOOK_CHANNEL:
+                  return mapper.readValue(msg.toString(), GateioOrderBookResponse.class);
+                case SPOT_TRADES_CHANNEL:
+                  return mapper.readValue(msg.toString(), GateioTradesResponse.class);
+              }
+              return mapper.readValue(msg.toString(), GateioWebSocketTransaction.class);
+            })
+        .filter(t -> currencyPair.equals(t.getCurrencyPair()));
+  }
+
+  public void subscribeMultipleCurrencyPairs(ProductSubscription... products) {
+    this.productSubscription = products[0];
+  }
+
+  public ProductSubscription getProduct() {
+    return this.productSubscription;
+  }
+
+  @Override
+  protected String getChannelNameFromMessage(JsonNode message) {
+    String channel = message.path("channel") != null ? message.path("channel").asText() : "";
+    String currencyPairOrderBook =
+        message.path("result").path("s") != null ? message.path("result").path("s").asText() : "";
+    String currencyPairTradesTickers =
+        message.path("result").path("currency_pair") != null
+            ? message.path("result").path("currency_pair").asText()
+            : "";
+
+    return new StringBuilder(channel)
+        .append(CHANNEL_NAME_DELIMITER)
+        .append(currencyPairOrderBook)
+        .append(currencyPairTradesTickers)
+        .toString();
+  }
+
+  @Override
+  public Observable<JsonNode> subscribeChannel(String channelName, Object... args) {
+    final CurrencyPair currencyPair =
+        (args.length > 0 && args[0] instanceof CurrencyPair) ? ((CurrencyPair) args[0]) : null;
+
+    String currencyPairChannelName =
+        String.format("%s-%s", channelName, currencyPair.toString().replace('/', '_'));
+
+    // Example channel name key: spot.order_book_update-ETH_USDT, spot.trades-BTC_USDT
+    if (!channels.containsKey(currencyPairChannelName)
+        && !subscriptions.containsKey(currencyPairChannelName)) {
+      subscriptions.put(
+          currencyPairChannelName, super.subscribeChannel(currencyPairChannelName, args));
+    }
+
+    return subscriptions.get(currencyPairChannelName);
+  }
+
+  /**
+   * Returns a JSON String containing the subscription message.
+   *
+   * @param channelName
+   * @param args CurrencyPair to subscribe to, followed by the refresh interval (Integer), followed
+   *     by the depth (Integer, only applicable for OrderBook channel subscriptions)
+   * @return
+   * @throws IOException
+   */
+  @Override
+  public String getSubscribeMessage(String channelName, Object... args) throws IOException {
+    final CurrencyPair currencyPair =
+        (args.length > 0 && args[0] instanceof CurrencyPair) ? ((CurrencyPair) args[0]) : null;
+    final Object[] argz =
```

Comment:
is there not a more elegant way to do this then nested args arr?

Reply:
I can't seem to sneak spaghetti code through you, I'll see if I can clean this up a bit.

Reply:
@9-9-0 don't you prefer a good pair of eyes ;)